### DISCUSSION TITLE
EntityRuler and NER pipe configuration

### DISCUSSION THREAD

**Discussion by cuiqui** [explosion/spaCy#10190] [Help: Coding & Implementations] [answered]
Hi all! First of all, I'm really loving Spacy 3. Now, let me describe the situation, the general idea is that given a text I want to recognize only one type of entity, so:

- I have an `EntityRuler` with lots of patters that I use to generate training data for the `NER`.
- Somewhere down the line I realized that the `EntityRuler` itself that I was using for training data generation was actually quite good for domain-specific sets of the general problem.
- It occurred to me that I could ship the model with two pipes then, first the `EntityRuler` would run and then the `NER` for cases in which some magic was needed to catch what the `EntityRuler` might have skipped because of its patterns. So in the configuration, the pipeline of the final production model looked like this: `["tok2vec", "entity_ruler", "ner"]`.

Now, the problem is that when I train with this configuration I'm using training data generated by the same `EntityRuler` that will run first in the pipeline. What I want is to have a domain-specific `EntityRuler` that I can use to generate training data for `NER` but that it does not take part in the training process of the final pipeline (`["tok2vec", "entity_ruler", "ner"]`), though I do want to include it in the model that I ship to production.

How can I approach this configuration-wise? Do I need to have two configs? One for the `EntityRuler`, and then another one that re-uses the `EntityRuler` generated with the first config, freezes the component, adds it to the pipeline before the `NER` and then trains the `NER`? How would this second configuration look like (the important fields)?

Am I approaching the problem from an extremely wrong angle? Thanks very much.

**Comment by polm** [comment, marked as answer]
It could make sense to have two configs, one for training and one for the full pipeline. [`spacy assemble`](https://spacy.io/api/cli#assemble) was made for this use case.

On the other hand, if you can include the initialized Entity Ruler at train time, you can just leave it there - if it's not an annotating component (as specified in the config) then the NER component won't see its annotations, so it just won't do anything. You can also mark it as a disabled component during training to keep it from running. This way you can use just one config and just override the `disabled_components` setting with your training command.

The [double NER sample project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double) might be a helpful reference, though it's combining multiple NER components.

**Reply by cuiqui** [reply]
Thanks very much! I noticed that my config has

```
[training]
annotating_components = []
```

So, as you mention, each component is trained in isolation and the Entity Ruler was running but not smearing the NER training. I will try to disable it as you suggest since there's no point in it running.

Going for the single config approach. I shouldn't mark the Entity Ruler as a frozen component since in that case it will exclude it from initialization, right? And I need it initialized so it's shipped with its patterns?

**Reply by polm** [reply]
"Frozen" means that the component is run during training but the weights in it aren't updated. There is no concept of "freezing" outside of training, and the EntityRuler has no weights anyway so freezing isn't relevant to it. So yes, you don't need to freeze it.
